### PR TITLE
Added CatBoostScorer

### PR DESCRIPTION
**Pull request by nikiduki**
В `catboost_scorer.py` добавил `dump` и `load` для CatBoostScorer. После можно будет их отредактировать и взять в `_dump_tools`

### CODE REVIEW

File: autointent/modules/scoring/_catboost/catboost_scorer.py [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+        self._model: CatBoostClassifier
+        self._embedder: Any
+        self._tokenizer: Tokenizer
+        self._dictionary: Dictionary
+
+    @classmethod
+    def from_context(
+        cls,
+        context: Context,
+        classification_model_config: EmbedderConfig | str | dict[str, Any] | None = None,
```
Comment (voorhs) [inline autointent/modules/scoring/_catboost/catboost_scorer.py]: лучше назвать `embedder_config`

File: autointent/modules/scoring/_catboost/catboost_scorer.py [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+
+    def get_classification_model_config(self) -> dict[str, Any]:
+        return self.classification_model_config.model_dump()
+
+    def get_implicit_initialization_params(self) -> dict[str, Any]:
+        return {
+            "classification_model_config": self.classification_model_config.model_dump(),
+        }
+
+    def _load_embedder(self) -> Any:  # noqa: ANN401
```
Comment (voorhs) [inline autointent/modules/scoring/_catboost/catboost_scorer.py]: у нас в AutoIntent реализована удобная обертка для эмбедера, посмотри как она используется в классе LinearScorer

File: autointent/modules/scoring/_catboost/catboost_scorer.py
Comment (voorhs) [inline autointent/modules/scoring/_catboost/catboost_scorer.py:1]: по-моему ты сделал что-то совсем не то :(

я поискал минут 30 и нашел вот такие ссылки:
- подробный туториал с кучей примеров: https://github.com/catboost/tutorials/blob/master/text_features/text_features_in_catboost.ipynb
- дока катбуста по текстовым фичам: https://catboost.ai/docs/en/features/text-features#additional-classes
- еще дока по текстовым фичам (тут объяснение как работает под капотом): https://catboost.ai/docs/en/concepts/algorithm-main-stages_text-to-numeric
- дока катбуста по эмбединг фичам: https://catboost.ai/docs/en/features/embeddings-features
- еще дока по эмбединг фичам (объяснение как работает под капотом): https://catboost.ai/docs/en/concepts/algorithm-main-stages_embedding-to-numeric

**Review by voorhs — CHANGES_REQUESTED** [review]

File: autointent/modules/scoring/_catboost/catboost_scorer.py [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+        self,
+        classification_model_config: EmbedderConfig | str | dict[str, Any] | None = None,
+        iterations: int = 100,
+        learning_rate: float = 0.1,
+        loss_function: str | None = None,
+        random_seed: int = 0,
+        verbose: bool = False,
+        **catboost_kwargs: Any,  # noqa: ANN401
+    ) -> None:
+        self.classification_model_config = EmbedderConfig.from_search_config(classification_model_config)
```
Comment (Samoed) [inline autointent/modules/scoring/_catboost/catboost_scorer.py]: Если `classification_model_config` `None`, то создастся `EmbedderConfig` с дефолтными значениями. У тебя тест `test_catboost_without_embedder` получается не то тестирует

File: autointent/modules/scoring/_catboost/catboost_scorer.py [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+        raw_max = getattr(tokenizer, "model_max_length", None)
+        max_len = (
+            DEFAULT_TOKEN_LENGTH
+            if not isinstance(raw_max, int) or raw_max <= 0 or raw_max > MAX_TOKEN_LENGTH
+            else raw_max
+        )
+
+        def encode(texts: list[str]) -> npt.NDArray[np.float32]:
+            with torch.no_grad():
+                batch = tokenizer(
```
Comment (Samoed) [inline autointent/modules/scoring/_catboost/catboost_scorer.py]: У `EmbedderConfig` есть параметры для токенизации в `tokenizer_config`

File: autointent/modules/scoring/_catboost/catboost_scorer.py [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+                return np.array(embeddings, dtype=np.float32)
+
+        self._embedder = encode
+        return self._embedder
+
+    def _init_text_tools(self) -> None:
+        if not hasattr(self, "_tokenizer"):
+            self._tokenizer = Tokenizer(lowercasing=True, separator_type="BySense", token_types=["Word", "Number"])
+        if not hasattr(self, "_dictionary"):
+            self._dictionary = Dictionary(occurence_lower_bound=1, gram_order=1)
```
Comment (Samoed) [inline autointent/modules/scoring/_catboost/catboost_scorer.py]: Тут тоже можно убрать

File: autointent/modules/scoring/_catboost/catboost_scorer.py [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+        self._validate_task(labels)
+
+        x = self._encode_utterances(utterances)
+        y: npt.NDArray[np.float32] | npt.NDArray[np.int64]
+        if self._multilabel:
+            y_mat = np.zeros((len(labels), self._n_classes), dtype=np.float32)
+            for i, lbls in enumerate(cast("Sequence[Sequence[int]]", labels)):
+                for class_i, lbl in enumerate(lbls):
+                    y_mat[i, class_i] = lbl
+            y = y_mat
```
Comment (Samoed) [inline autointent/modules/scoring/_catboost/catboost_scorer.py]: Можно просто заменить на np.asarray(labels)?

**Review by Samoed — CHANGES_REQUESTED** [review]

File: autointent/modules/scoring/_catboost/catboost_scorer.py [hunk 7]
```diff
@@ -122,68 +118,29 @@ def from_context(
...
+    def _init_catboost_text_tools(self) -> None:
         if not hasattr(self, "_tokenizer"):
             self._tokenizer = Tokenizer(lowercasing=True, separator_type="BySense", token_types=["Word", "Number"])
         if not hasattr(self, "_dictionary"):
             self._dictionary = Dictionary(occurence_lower_bound=1, gram_order=1)
+        if not hasattr(self, "_dictionary_fitted"):
+            self._dictionary_fitted = False
+
+    def get_embedder_config(self) -> dict[str, Any]:
+        return self.embedder_config.model_dump()
```
Comment (Samoed) [inline autointent/modules/scoring/_catboost/catboost_scorer.py]: Если `self._use_embedder` False, то ошибка будет

File: autointent/modules/scoring/_catboost/catboost_scorer.py [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+        )
+
+
+    def _init_catboost_text_tools(self) -> None:
+        if not hasattr(self, "_tokenizer"):
+            self._tokenizer = Tokenizer(lowercasing=True, separator_type="BySense", token_types=["Word", "Number"])
+        if not hasattr(self, "_dictionary"):
+            self._dictionary = Dictionary(occurence_lower_bound=1, gram_order=1)
+        if not hasattr(self, "_dictionary_fitted"):
+            self._dictionary_fitted = False
```
Comment (Samoed) [inline autointent/modules/scoring/_catboost/catboost_scorer.py]: Надо сделать как-то задаваемым как так? https://catboost.ai/docs/en/references/text-processing__test-processing__default-value
Просто кажется что можно не указывать tokenizer, dictironary и тд

**Review by Samoed — APPROVED** [review]
